### PR TITLE
new(drag): add `useDrag` hook

### PR DESCRIPTION
**Pull request by williaster**
#### :rocket: Enhancements

This PR 
- adds a `useDrag` hook to `@visx/drag` to complement the render props `<Drag />` component 
  - **(I need this for some `@visx/annotation` work)**
- refactors `<Drag />` to use `useDrag`
- bumps `react` `peerDependency` to `^16.8.0-0` to support hooks
- adds support for `PointerEvents` (in addition to `Mouse` and `Touch` events) in drag callbacks
- updates `@visx/demo` `/drag-ii` to use `useDrag` so we have examples of both usages
- updates `/docs/drag` to include `useDrag`
- adds minimal tests for all `@visx/drag` apis

Tested via
- verified `/drag-ii` is still functional
- Updated docs

@hshoff @kristw

### CODE REVIEW

File: packages/visx-demo/src/sandboxes/visx-drag-i/Example.tsx [hunk 1]
```diff
@@ -64,27 +64,25 @@ export default function DragI({ width, height }: DragIProps) {
               setDraggingItems(raise(draggingItems, i));
             }}
           >
-            {({ dragStart, dragEnd, dragMove, isDragging, dx, dy }) =>
-              (false && isDragging && console.log(d.id, d.x, dx)) || (
```
Comment (williaster) [inline packages/visx-demo/src/sandboxes/visx-drag-i/Example.tsx:68]: just removed this guy 😱

File: packages/visx-demo/src/sandboxes/visx-drag-ii/Example.tsx [hunk 2]
```diff
@@ -33,63 +58,45 @@ export default function DragII({ data = [], width, height }: DragIIProps) {
...
-              nextLines[lastIndex] = [...(nextLines[lastIndex] || []), newPoint];
-              return nextLines;
-            });
-          }}
-        >
-          {({ x = 0, y = 0, dx, dy, isDragging, dragStart, dragEnd, dragMove }) => (
+
+        <g>
+          {isDragging && (
+            /* capture mouse events (note: <Drag /> does this for you) */
```
Comment (williaster) [inline packages/visx-demo/src/sandboxes/visx-drag-ii/Example.tsx:64]: 💬  this is the key missing feature from `useDrag`

File: packages/visx-drag/src/util/useStateWithCallback.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  const setStateCallback = useCallback<SetStateWithCallback<State>>(
+    (nextState, callback) => {
+      callbackRef.current = callback || null;
+      setState(nextState);
+    },
+    [setState],
+  );
+
+  // if we use useEffect, some callback invocations are skipped
+  useLayoutEffect(() => {
```
Comment (williaster) [inline packages/visx-drag/src/util/useStateWithCallback.ts:24]: the hook is the general approach I saw for trying to re-create the `setState(state, callback)` component api in hooks, except they all recommend `useEffect`.  However, when I used `useEffect` I noticed in the `/drag-ii` demo that some of the callbacks were skipped (e.g., the `useDragOptions.onDragStart` [the callback] would be skipped sometimes, which resulted in the previous line being connected to the new line you were starting to draw [`onDragStart` creates a new line in `/drag-ii`]).

I don't have a ton of experience using `useLayoutEffect` vs `useEffect`, I just know that `useLayoutEffect` is blocking while `useEffect` isn't as documented in the react api and we get a warning in `next` now that `useLayoutEffect` cannot be used server-side (for this functionality I don't think it matters).  

Would love any thoughts on this!

File: packages/visx-drag/src/useDrag.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  dragStart: (event: MouseOrTouchEvent) => void;
+};
+
+/** Hook for dragging, returns a `UseDrag` object. */
+export default function useDrag({
+  resetOnStart = false,
+  onDragEnd,
+  onDragMove,
+  onDragStart,
+}: UseDragOptions): UseDrag {
```
Comment (williaster) [inline packages/visx-drag/src/useDrag.tsx:51]: note to make this optional to enable `useDrag()` if desired

File: packages/visx-drag/src/useDrag.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
+import React, { useCallback } from 'react';
+import { localPoint } from '@visx/event';
+import useStateWithCallback from './util/useStateWithCallback';
+
+type MouseOrTouchEvent = React.MouseEvent | React.TouchEvent;
```
Comment (hshoff) [inline packages/visx-drag/src/useDrag.tsx:5]: Could add `React.PointerEvent`. Was added in React 16.4 https://reactjs.org/blog/2018/05/23/react-v-16-4.html#pointer-events

File: packages/visx-drag/src/util/useStateWithCallback.ts (hunk 3 above)
Comment (hshoff) [reply packages/visx-drag/src/util/useStateWithCallback.ts:24]: Hm. Not ideal but this could be something we revisit in the future.

**Review by hshoff — APPROVED** [review]

**Review by hshoff — CHANGES_REQUESTED** [review]
Just remembered, need to bump react peerDep to `^16.8.0-0`. similar to [@visx/tooltip](https://github.com/airbnb/visx/blob/master/packages/visx-tooltip/package.json#L39).  (hooks were added in 16.8 https://reactjs.org/docs/hooks-intro.html). 
https://github.com/airbnb/visx/blob/master/packages/visx-tooltip/package.json#L39

**Review by hshoff — APPROVED** [review]

**Pull request merged**